README.md
# Apache Kafka 0.8+ protocol for Erlang

## Summary

This is a very early implementation and provides only
some of the basic features of the Kafka protocol:

* synchronous and asynchronous requests;
* produce - posting data to the Kafka;
* fetch - consuming data from the Kafka;
* offsets - lookup available offsets for particular topic and partition;
* metadata - lookup available Kafka nodes and topics;
* offset commit - save consumer current position;
* offset fetch - retrieve consumer position from Kafka.

The goal of the project is to provide comprehensive Erlang bindings to
the Apache Kafka protocol as is, without any client-side extensions. So,
you should keep
[Apache Kafka protocol description]
(https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol)
page open when you use the library. As advantage of such approach:

* the library can be easily extended with support of new request/response types
without any changes in the API;
* the code required to add support of a new request/response is extremely small.
For example, here is the code from kafka_api module which add support for
OffsetCommit request/response API:

```erlang
...
req_payload(?OffsetCommit, 0 = _ApiVersion) ->
    ?Struct(
       [string, %% group_id
        ?ArrayOf( %% topics
           ?Struct(
              [string, %% topic_name
               ?ArrayOf( %% partitions
                  ?Struct(
                     [int32, %% partition_id
                      int64, %% offset
                      string %% metadata
                     ]))
              ]))
       ]);
...
resp_payload(?OffsetCommit, 0 = _ApiVersion) ->
    ?ArrayOf( %% responses
       ?Struct(
          [string, %% topic_name
           ?ArrayOf( %% partition_responses
              ?Struct(
                 [int32, %% partition_id
                  int16 %% error_code
                 ]))
          ]));
...
```

So, if you looking for high level client library with bells and whistles, you
should better use:

* [ekaf](https://github.com/helpshift/ekaf) - Kafka producer library with buffering,
 batch send and other convenient features.
* [brod](https://github.com/klarna/brod) - Kafka client which provide nice
 Erlang MFAs to make several commonly used Kafka requests.

But if you want to build you own logic based on bare Apache Kafka protocol - you
are on the right way.

## License

BSD two-clause license.

## Examples

### Start Kafka client as linked process:

```erlang
Host = "10.0.0.1",
Port = 9092,
{ok, Socket} = kafka_socket:start_link(Host, Port, []),
...
```

The example code snippet above will crash on connection error. There is a
special option for kafka_socket:start_link/3 function which make the
process to behave more stable in case of network failures:

```erlang
{ok, Socket} = kafka_socket:start_link("10.0.0.1", 9092, [reconnect]),
...
```

In this case the process will not terminate until kafka_socket:close/1
will be called and will try to continuously reconnect to the broker if
first connection attempt was failed or when already established connection
closed for some reason.

### Request metadata from the broker:

```erlang
case kafka_socket:sync(Socket,
                       _ApiKey = 3,
                       _ApiVersion = 0,
                       _ClientID = "my-client",
                       _MetadataRequest = [],
                       _SockReadTimeout = infinity) of
    {ok, {Brokers, Topics} = _Metadata} ->
        [... || {BrokerID, Host, Port} <- Brokers],
        [... || {TopicErrorCode, TopicName, Partitions} <- Topics,
                {PartitionErrorCode, PartitionID, LeaderBrokerID,
                 ReplicaBrokerIDs, InSyncReplicaBrokerIDs} <- Partitions],
        ...;
    {error, Reason} ->
        ...
end,
```

or use an existing bootstrap wrapper which will connect, make a
metadata request and disconnect:

```erlang
case kafka:metadata(_Host = "10.0.0.1", _Port = 9092) of
    {ok, Brokers, Topics} ->
        [... || {BrokerID, Host, Port} <- Brokers],
        [... || {TopicErrorCode, TopicName, Partitions} <- Topics,
                {PartitionErrorCode, PartitionID, LeaderBrokerID,
                 ReplicaBrokerIDs, InSyncReplicaBrokerIDs} <- Partitions],
        ...;
    {error, Reason} ->
        ...
end,
```

### Make a produce request:

```erlang
Topics =
    [{_TopicName = "my-topic",
      _Partitions =
          [{_PartitionID = 0,
            _MessageSet =
                [{_AutoOffset = -1,
                  _Message =
                      {_MagicByte = 1,
                       _Attributes = 0,
                       _Key = <<"k2">>,
                       _Value = <<"v2">>}}]}]}],
ProduceRequest = {_AckAfterLocalWrite = 1, _TimeoutMillis = 1000, Topics},
case kafka_socket:sync(Socket,
                       _ApiKey = 3,
                       _ApiVersion = 0,
                       _ClientID = "my-client",
                       ProduceRequest,
                       _SockReadTimeout = infinity) of
    {ok, [{_TopicName, Partitions}]} ->
        [... || {PartitionID, ErrorCode, BaseOffset} <- Partitions],
        ...;
    {error, Reason} ->
        ...
end,
```

### Make an asynchronous produce request:

```erlang
Topics =
    [{_TopicName = "my-topic",
      _Partitions =
          [{_PartitionID = 0,
            _MessageSet =
                [{_AutoOffset = -1,
                  _Message =
                      {_MagicByte = 1,
                       _Attributes = 0,
                       _Key = <<"k2">>,
                       _Value = <<"v2">>}}]}]}],
ProduceRequest = {_NoAcknowledge = 0, _TimeoutMillis = 1000, Topics},
case kafka_socket:async(Socket,
                        _ApiKey = 3,
                        _ApiVersion = 0,
                        _ClientID = "my-client",
                        ProduceRequest) of
    ok ->
        ...;
    {error, Reason} ->
        ...
end,
```

### Make a fetch (consume) request:

```erlang
Topics =
    [{_TopicName = "my-topic",
      _Partitions = [{_PartitionID = 0,
                      _FetchOffset = 0,
                      _MaxBytes = 1024}]}],
FetchRequest = {_ReplicaID = -1, _MaxWaitTimeMillis = 1000, _MinBytes = 1, Topics},
case kafka_socket:sync(Socket,
                       _ApiKey = 1,
                       _ApiVersion = 0,
                       _ClientID = "my-client",
                       FetchRequest,
                       _SockReadTimeout = infinity) of
    {ok, TopicReplies} ->
        [... || {TopicName, Partitions} <- TopicReplies,
                {PartitionID, ErrorCode, HighWatermark, MessageSet} <- Partitions,
                {Offset, {MagicByte, Attrs, Key, Value}} <- MessageSet],
        ...;
    {error, Reason} ->
        ...
end,
```

### Make an offsets request:

```erlang
OffsetsRequest =
    {_ReplicaID = -1,
     _Topics = [{_TopicName = "my-topic",
                 _Partitions = [{_PartitionID = 0,
                                 _Timestamp = -1, %% special value means 'latest offset'
                                 _MaxNumOffsets = 1}]}]},
case kafka_socket:sync(Socket,
                       _ApiKey = 2,
                       _ApiVersion = 0,
                       _ClientID = "my-client",
                       OffsetsRequest,
                       _SockReadTimeout = infinity) of
    {ok, TopicReplies} ->
        [... || {TopicName, Partitions} <- TopicReplies,
                {PartitionID, ErrorCode, Offsets} <- Partitions,
                Offset <- Offsets],
        ...;
    {error, Reason} ->
        ...
end,
```

### Make an offset commit request:

```erlang
OffsetCommitRequest =
    {_GroupID = "my-consumer-id",
     _Topics = [{_TopicName = "my-topic",
                 _Partitions = [{_PartitionID = 0,
                                 _Offset = 1234,
                                 _Metadata = "any data"}]}]},
case kafka_socket:sync(Socket,
                       _ApiKey = 8,
                       _ApiVersion = 0,
                       _ClientID = "my-client",
                       OffsetCommitRequest,
                       _SockReadTimeout = infinity) of
    {ok, TopicReplies} ->
        [... || {TopicName, Partitions} <- TopicReplies,
                {PartitionID, ErrorCode} <- Partitions],
        ...;
    {error, Reason} ->
        ...
end,
```

### Make an offset fetch request:

```erlang
OffsetFetchRequest =
    {_GroupID = "my-consumer-id",
     _Topics = [{_TopicName = "my-topic",
                 _Partitions = [0, 1, 2]}]},
case kafka_socket:sync(Socket,
                       _ApiKey = 9,
                       _ApiVersion = 0,
                       _ClientID = "my-client",
                       OffsetFetchRequest,
                       _SockReadTimeout = infinity) of
    {ok, TopicReplies} ->
        [... || {TopicName, Partitions} <- TopicReplies,
                {PartitionID, Offset, Metadata, ErrorCode} <- Partitions],
        ...;
    {error, Reason} ->
        ...
end,
```

## Build dependencies

* GNU Make;
* Erlang OTP;
* erlang-dev, erlang-tools (only when they are not provided with the Erlang OTP, e.g. in Debian);
* erlang-edoc (optional, needed to generate HTML docs);
* erlang-dialyzer (optional, needed to run the Dialyzer).

## Runtime dependencies

* Erlang OTP.

## Build

```make compile```

## Generate an HTML documentation for the code

```make html```

## Dialyze the code

```make dialyze```

## TODO

* implement message set compression;
* add codec specs for most of Apache Kafka request/response pairs.


debian/changelog
erlang-kafka (1.4.0) unstable; urgency=low

  * consumer: extend API for reconsume/2 function
  * consumer: add set_max_bytes/2 API call

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Fri, 01 Jul 2016 15:57:18 +0300

erlang-kafka (1.3.0) unstable; urgency=low

  * consumer: simplify suspend/resume schema
  * consumer: add resume/1 API call
  * consumer: add unread_messages/1 API call
  * consumer: implement status change notifications
  * _d API: minor fixes

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Sun, 05 Jun 2016 15:23:56 +0300

erlang-kafka (1.2.0) unstable; urgency=low

  * consumer: add flow control

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Fri, 03 Jun 2016 09:53:55 +0300

erlang-kafka (1.1.1) unstable; urgency=low

  * consumer: fix function spec
  * consumer: trap exits
  * make crd/7 call more user friendly
  * kafka_d:produce/5 and examples minor fix

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Wed, 01 Jun 2016 19:43:30 +0300

erlang-kafka (1.1.0) unstable; urgency=low

  * implement high level asynchronous consumer
  * add module for simple wrappers
  * _socket: add extra tracing
  * macros: add missing 1 ErrorCode
  * add human readable mapping for error codes
  * codec: allow Kafka "string" to be Erlang binary()
  * add metadata/1 convenient wrapper in the main module
  * define some commonly used data types
  * _socket: add new option - sync_start
  * _socket: minor fix - add missing "behaviour" module option
  * _socket: implement {state_listener,X} option
  * deb specs: run unit tests during DEB package build
  * deb specs: pack header files in erlang-kafka-dev DEB package
  * deb specs: minor description fix

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Tue, 05 Apr 2016 15:58:41 +0300

erlang-kafka (1.0.0) unstable; urgency=low

  * MAJOR: Total refactor. Less code, more features
  * minor macro names fix
  * _proto: shorten utility functions, cover with tests

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Mon, 04 Apr 2016 15:22:44 +0300

erlang-kafka (0.2.3) unstable; urgency=low

  * deb specs: simplify erlang app staged installation
  * deb specs: simplify docs staged installation
  * dispose of version file

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Thu, 31 Mar 2016 19:52:43 +0300

erlang-kafka (0.2.2) unstable; urgency=low

  * fix crash while reconnection;
  * decrease reconnect periods;
  * reconnect when no brokers available;
  * extend readme with code examples.

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Mon, 26 Jan 2015 17:26:00 +0200

erlang-kafka (0.2.0-1) unstable; urgency=low

  * Initial release.

 -- Aleksey Morarash <aleksey.morarash@gmail.com>  Thu, 07 Aug 2014 11:36:00 +0300


src/kafka.erl
%%% @doc
%%% Erlang Client for the Apache Kafka 0.8+.
%%%
%%% Unlike most Erlang libraries, this main module contains almost
%%% nothing interesting.
%%% To manage connection to the Kafka broker and issue requests
%%% use kafka_socket module instead.

%%% @author Aleksey Morarash <aleksey.morarash@gmail.com>
%%% @since 04 Aug 2014
%%% @copyright 2014, Aleksey Morarash <aleksey.morarash@gmail.com>

-module(kafka).

%% API exports
-export(
   [metadata/2,
    metadata/1,
    crd/7
   ]).

-include("kafka.hrl").
-include("kafka_types.hrl").
-include("kafka_constants.hrl").

%% these types are defined in include/kafka_types.hrl
-export_types(
   [int8/0,
    int16/0,
    int32/0,
    int64/0,
    kstring/0,
    kbytes/0,
    topic_name/0,
    partition_id/0,
    offset/0
   ]).

%% --------------------------------------------------------------------
%% API functions
%% --------------------------------------------------------------------

%% @doc Connect to broker, make a metadata request for all existing
%% topics and disconnect.
%% This function is a convenient wrapper to make a bootstrapping
%% metadata request to a Kafka broker.
-spec metadata(Host :: kafka_socket:host(),
               Port :: inet:port_number()) ->
                      {ok, Brokers :: list(), Topics :: list()} |
                      {error, Reason :: any()}.
metadata(Host, Port) ->
    case crd(Host, Port, ?Metadata, _ApiVersion = 0,
             _ClientID = "erlang-kafka", _TopicsToFetch = [],
             _ReadSockTimeout = infinity) of
        {ok, {Brokers, Topics}} ->
            {ok, Brokers, Topics};
        {error, _Reason} = Error ->
            Error
    end.

%% @doc Fetch metadata from first available Kafka broker.
%% This function is a convenient wrapper to make a bootstrapping
%% metadata request to a Kafka broker.
-spec metadata([{kafka_socket:host(), inet:port_number()}]) ->
                      {Brokers :: list(), Topics :: list()} |
                      undefined.
metadata([]) ->
    undefined;
metadata([{Host, Port} | Tail]) ->
    case metadata(Host, Port) of
        {ok, [] = _Brokers, _Topics} ->
            %% This strange behaviour I watched with kafka 0.8.2.2.
            %% Try next node.
            metadata(Tail);
        {ok, Brokers, Topics} ->
            {Brokers, Topics};
        {error, _Reason} ->
            metadata(Tail)
    end.

%% @doc Connect to broker, make a request and disconnect.
%% This function is created mostly for debugging purposes.
%% Unlikely you want to establish new TCP connection on each subsequent
%% produce/consume/etc request in production system.
-spec crd(Host :: kafka_socket:host(),
          Port :: inet:port_number(),
          ApiKey :: int16(),
          ApiVersion :: int16(),
          ClientID :: kstring(),
          RequestPayload :: any(),
          Timeout :: timeout()) ->
                 {ok, ResponsePayload :: any()} | {error, Reason :: any()}.
crd(Host, Port, ApiKey, ApiVersion, ClientID, RequestPayload, Timeout) ->
    case kafka_socket:start_link(Host, Port, []) of
        {ok, Socket} ->
            Result = kafka_socket:sync(Socket, ApiKey, ApiVersion, ClientID,
                                       RequestPayload, Timeout),
            ok = kafka_socket:close(Socket),
            Result;
        {error, _Reason} = Error ->
            Error
    end.


src/kafka_api.erl
%%% @doc
%%% Apache Kafka API specifications.
%%%
%%% The module contains request/response message specifications,
%%% used to encode/decode the messages for network transmission.
%%%
%%% For more details about specifications see kafka_codec module.

%%% @author Aleksey Morarash <aleksey.morarash@gmail.com>
%%% @since 02 Apr 2016
%%% @copyright 2016, Aleksey Morarash <aleksey.morarash@gmail.com>

-module(kafka_api).

%% API exports
-export([req/2, resp/2, message_set/1]).

-include("kafka.hrl").
-include("kafka_types.hrl").
-include("kafka_constants.hrl").

-define(ArrayOf(Type), [Type]).
-define(Struct(ElemTypes), {ElemTypes}).

%% ----------------------------------------------------------------------
%% API functions
%% ----------------------------------------------------------------------

%% @doc Return request message specification for encoding/decoding.
-spec req(ApiKey :: int16(), ApiVersion :: int16()) -> kafka_codec:ktype().
req(ApiKey, ApiVersion) ->
    ?Struct(
       [int16, %% api_key
        int16, %% api_version
        int32, %% corellation_id
        string, %% client_id
        req_payload(ApiKey, ApiVersion) %% request message payload
       ]).

%% @doc Return response message specification for encoding/decoding.
-spec resp(ApiKey :: int16(), ApiVersion :: int16()) -> kafka_codec:ktype().
resp(ApiKey, ApiVersion) ->
    ?Struct(
       [int32, %% corellation_id
        resp_payload(ApiKey, ApiVersion) %% response message payload
       ]).

%% @doc Return codec specification for MessageSet structure.
%% This part of the request/response messages is exported because
%% message sets can be recursively nested (when compression is applied).
%% So we allow user to play with message sets by himself.
-spec message_set(ApiVersion :: int16()) -> kafka_codec:ktype().
message_set(0) ->
    {mset,
     ?Struct(
        [int64, %% offset
         {msg,
          ?Struct(
             [int8, %% magic byte
              int8, %% attributes
              bytes, %% key
              bytes %% value
             ])}
        ])};
message_set(1) ->
    {mset,
     ?Struct(
        [int64, %% offset
         {msg,
          ?Struct(
             [int8, %% magic byte
              int8, %% attributes
              int64, %% timestamp
              bytes, %% key
              bytes %% value
             ])}
        ])}.

%% ----------------------------------------------------------------------
%% Internal functions
%% ----------------------------------------------------------------------

%% @doc Return specification for payload of request message.
-spec req_payload(ApiKey :: int16(), ApiVersion :: int16()) ->
                         kafka_codec:ktype().
req_payload(?Produce, ApiVersion) ->
    ?Struct(
       [int16, %% acks
        int32, %% timeout
        ?ArrayOf( %% topic_data
           ?Struct(
              [string, %% topic_name
               ?ArrayOf( %% data
                  ?Struct(
                     [int32, %% partition
                      message_set(ApiVersion) %% record_set
                     ]))
              ]))
       ]);
req_payload(?Fetch, 0) ->
    ?Struct(
       [int32, %% replica_id
        int32, %% max_wait_time
        int32, %% min_bytes
        ?ArrayOf( %% topics
           ?Struct(
              [string, %% topic_name
               ?ArrayOf( %% partitions
                  ?Struct(
                     [int32, %% partition
                      int64, %% fetch_offset
                      int32 %% max_bytes
                     ]))
              ]))
       ]);
req_payload(?Fetch, 1) ->
    req_payload(?Fetch, 0);
req_payload(?Fetch, 2) ->
    req_payload(?Fetch, 1);
req_payload(?Offsets, 0) ->
    ?Struct(
       [int32, %% replica_id
        ?ArrayOf( %% topics
           ?Struct(
              [string, %% topic_name
               ?ArrayOf( %% partitions
                  ?Struct(
                     [int32, %% partition
                      int64, %% timestamp
                      int32 %% max_num_offsets
                     ]))
              ]))
       ]);
req_payload(?Offsets, 1) ->
    req_payload(?Offsets, 0);
req_payload(?Offsets, 2) ->
    req_payload(?Offsets, 1);
req_payload(?Metadata, 0) ->
    ?ArrayOf(string); %% topic_names
req_payload(?Metadata, 1) ->
    req_payload(?Metadata, 0);
req_payload(?Metadata, 2) ->
    req_payload(?Metadata, 1);
req_payload(?OffsetCommit, 0) ->
    ?Struct(
       [string, %% group_id
        ?ArrayOf( %% topics
           ?Struct(
              [string, %% topic_name
               ?ArrayOf( %% partitions
                  ?Struct(
                     [int32, %% partition_id
                      int64, %% offset
                      string %% metadata
                     ]))
              ]))
       ]);
req_payload(?OffsetCommit, 1) ->
    ?Struct(
       [string, %% group_id
        int32, %% group_generation_id
        string, %% member_id
        ?ArrayOf( %% topics
           ?Struct(
              [string, %% topic_name
               ?ArrayOf( %% partitions
                  ?Struct(
                     [int32, %% partition_id
                      int64, %% offset
                      int64, %% timestamp
                      string %% metadata
                     ]))
              ]))
       ]);
req_payload(?OffsetCommit, 2) ->
    ?Struct(
       [string, %% group_id
        int32, %% group_generation_id
        string, %% member_id
        int64, %% retention_time
        ?ArrayOf( %% topics
           ?Struct(
              [string, %% topic_name
               ?ArrayOf( %% partitions
                  ?Struct(
                     [int32, %% partition_id
                      int64, %% offset
                      int64, %% timestamp
                      string %% metadata
                     ]))
              ]))
       ]);
req_payload(?OffsetFetch, 0) ->
    ?Struct(
       [string, %% group_id
        ?ArrayOf( %% topics
           ?Struct(
              [string, %% topic_name
               ?ArrayOf(int32) %% partitions
              ]))
       ]);
req_payload(?OffsetFetch, 1) ->
    req_payload(?OffsetFetch, 0);
req_payload(?OffsetFetch, 2) ->
    req_payload(?OffsetFetch, 0).

%% @doc Return specification for payload of response message.
-spec resp_payload(ApiKey :: int16(), ApiVersion :: int16()) ->
                          kafka_codec:ktype().
resp_payload(?Produce, 0) ->
    ?ArrayOf(
       ?Struct(
          [string, %% topic_name
           ?ArrayOf( %% partitions_responses array
              ?Struct(
                 [int32, %% partition
                  int16, %% error_code
                  int64 %% base_offset
                 ]))
          ]));
resp_payload(?Produce, 1) ->
    ?Struct(
       [?ArrayOf( %% as for v0
           ?Struct(
              [string, %% topic_name
               ?ArrayOf( %% partitions_responses array
                  ?Struct(
                     [int32, %% partition
                      int16, %% error_code
                      int64 %% base_offset
                     ]))
              ])),
        int32 %% throttle_time_ms
       ]);
resp_payload(?Produce, 2) ->
    resp_payload(?Produce, 1);
resp_payload(?Fetch, 0) ->
    ?ArrayOf(
       ?Struct(
          [string, %% topic_name
           ?ArrayOf( %% partition_responses
              ?Struct(
                 [int32, %% partition_id
                  int16, %% error_code
                  int64, %% high_watermark
                  message_set(0)
                 ]))
          ]));
resp_payload(?Fetch, 1) ->
    ?Struct(
       [int32, %% throttle_time_ms
        ?ArrayOf( %% responses
           ?Struct(
              [string, %% topic_name
               ?ArrayOf( %% partition_responses
                  ?Struct(
                     [int32, %% partition_id
                      int16, %% error_code
                      int64, %% high_watermark
                      message_set(1)
                     ]))
              ]))]);
resp_payload(?Fetch, 2) ->
    resp_payload(?Fetch, 1);
resp_payload(?Offsets, 0) ->
    ?ArrayOf( %% responses
       ?Struct(
          [string, %% topic_name
           ?ArrayOf( %% partition_responses
              ?Struct(
                 [int32, %% partition_id
                  int16, %% error_code
                  ?ArrayOf(int64) %% offsets
                 ]))
          ]));
resp_payload(?Offsets, 1) ->
    resp_payload(?Offsets, 0);
resp_payload(?Offsets, 2) ->
    resp_payload(?Offsets, 1);
resp_payload(?Metadata, 0) ->
    ?Struct(
       [?ArrayOf( %% brokers
           ?Struct(
              [int32, %% node_id
               string, %% host
               int32 %% port
              ])
          ),
        ?ArrayOf( %% topic_metadata
           ?Struct(
              [int16, %% topic_error_code
               string, %% topic_name
               ?ArrayOf(
                  ?Struct(
                     [int16, %% partition_error_code
                      int32, %% partition_id
                      int32, %% leader_node_id
                      ?ArrayOf(int32), %% replica node ids
                      ?ArrayOf(int32) %% in-sync-with-leader replica node ids
                     ]))
              ])
          )]);
resp_payload(?Metadata, 1) ->
    resp_payload(?Metadata, 0);
resp_payload(?Metadata, 2) ->
    resp_payload(?Metadata, 1);
resp_payload(?OffsetCommit, 0) ->
    ?ArrayOf( %% responses
       ?Struct(
          [string, %% topic_name
           ?ArrayOf( %% partition_responses
              ?Struct(
                 [int32, %% partition_id
                  int16 %% error_code
                 ]))
          ]));
resp_payload(?OffsetCommit, 1) ->
    resp_payload(?OffsetCommit, 0);
resp_payload(?OffsetCommit, 2) ->
    resp_payload(?OffsetCommit, 1);
resp_payload(?OffsetFetch, 0) ->
    ?ArrayOf( %% responses
       ?Struct(
          [string, %% topic_name
           ?ArrayOf( %% partition_responses
              ?Struct(
                 [int32, %% partition_id
                  int64, %% offset
                  string, %% metadata
                  int16 %% error_code
                 ]))
          ]));
resp_payload(?OffsetFetch, 1) ->
    resp_payload(?OffsetFetch, 0);
resp_payload(?OffsetFetch, 2) ->
    resp_payload(?OffsetFetch, 1).


src/kafka_d.erl
%%% @doc
%%% Apache Kafka API "for dummies".
%%%
%%% This module provides simplified wrappers for some useful
%%% requests to the Kafka broker.
%%%
%%% First argument of each function is a PID of socket process
%%% created by kafka_socket:start_link/3 function.
%%%
%%% I wrote this stuff in a separate module because all
%%% this functions are mostly version specific and may crash
%%% when Kafka protocol change.
%%% The module is compromise between lowlevel API provided by
%%% kafka_socket module and highlevel features usually provided
%%% by mature Kafka clients.
%%%
%%% I'm certain will not implement all possible and/or impossible
%%% wrappers here because Apache Kafka protocol is tricky and
%%% it can map to an infinite amount of "simple wrappers".

%%% @author Aleksey Morarash <aleksey.morarash@gmail.com>
%%% @since 04 Apr 2016
%%% @copyright 2016, Aleksey Morarash <aleksey.morarash@gmail.com>

-module(kafka_d).

%% API exports
-export(
   [get_first_offset/3,
    get_last_offset/3,
    get_current_offset/4,
    set_current_offset/5,
    produce/5
   ]).

-include("kafka.hrl").
-include("kafka_types.hrl").
-include("kafka_constants.hrl").

%% --------------------------------------------------------------------
%% API functions
%% --------------------------------------------------------------------

%% @doc Fetch the very first available offset of the partition.
-spec get_first_offset(Socket :: pid(), topic_name(), partition_id()) ->
                              {ok, Offset :: offset()} |
                              {error, Reason :: any()}.
get_first_offset(Socket, Topic0, Partition) ->
    Topic = canonicalize_string(Topic0),
    case kafka_socket:sync(
           Socket, ?Offsets, _ApiVersion = 0, _ClientID = undefined,
           {_ReplicaID = -1, [{Topic, [{Partition, -2, _Len = 1}]}]},
           _SockReadTimeout = infinity) of
        {ok, [{Topic, [{Partition, ?NONE = _ErrorCode, [Offset]}]}]} ->
            {ok, Offset};
        {ok, [{Topic, [{Partition, ErrorCode, _}]}]} ->
            {error, ?err2atom(ErrorCode)};
        {error, _Reason} = Error ->
            Error
    end.

%% @doc Fetch the latest available offset of the partition.
-spec get_last_offset(Socket :: pid(), topic_name(), partition_id()) ->
                             {ok, Offset :: offset()} |
                             {error, Reason :: any()}.
get_last_offset(Socket, Topic0, Partition) ->
    Topic = canonicalize_string(Topic0),
    case kafka_socket:sync(
           Socket, ?Offsets, _ApiVersion = 0, _ClientID = undefined,
           {_ReplicaID = -1, [{Topic, [{Partition, -1, _Len = 1}]}]},
           _SockReadTimeout = infinity) of
        {ok, [{Topic, [{Partition, ?NONE = _ErrorCode, [Offset]}]}]} ->
            {ok, Offset};
        {ok, [{Topic, [{Partition, ErrorCode, _}]}]} ->
            {error, ?err2atom(ErrorCode)};
        {error, _Reason} = Error ->
            Error
    end.

%% @doc Fetch stored offset for GroupID in partition.
-spec get_current_offset(Socket :: pid(),
                         topic_name(),
                         partition_id(),
                         group_id()) ->
                                {ok, Offset :: offset()} |
                                {error, Reason :: any()}.
get_current_offset(Socket, Topic0, Partition, GroupID) ->
    Topic = canonicalize_string(Topic0),
    case kafka_socket:sync(
           Socket, ?OffsetFetch, _ApiVersion = 0, _ClientID = undefined,
           {GroupID, [{Topic, [Partition]}]},
           _SockReadTimeout = infinity) of
        {ok, [{Topic, [{Partition, Offset, _Metadata, ?NONE = _ErrorCode}]}]} ->
            {ok, Offset};
        {ok, [{Topic, [{Partition, _, _, ErrorCode}]}]} ->
            {error, ?err2atom(ErrorCode)};
        {error, _Reason} = Error ->
            Error
    end.

%% @doc Store offset for GroupID in partition.
-spec set_current_offset(Socket :: pid(),
                         topic_name(),
                         partition_id(),
                         group_id(),
                         offset()) ->
                                ok | {error, Reason :: any()}.
set_current_offset(Socket, Topic0, Partition, GroupID, Offset) ->
    Topic = canonicalize_string(Topic0),
    case kafka_socket:sync(
           Socket, ?OffsetCommit, _ApiVersion = 0, _ClientID = undefined,
           {GroupID, [{Topic, [{Partition, Offset, _Metadata = undefined}]}]},
           _SockReadTimeout = infinity) of
        {ok, [{_Topic, [{Partition, ?NONE = _ErrorCode}]}]} ->
            ok;
        {ok, [{_Topic, [{Partition, ErrorCode}]}]} ->
            {error, ?err2atom(ErrorCode)};
        {error, _Reason} = Error ->
            Error
    end.

%% @doc Post a single message to the partition.
-spec produce(Socket :: pid(),
              topic_name(),
              partition_id(),
              Key :: kafka:kbytes(),
              Value :: kafka:kbytes()) ->
                     ok | {error, Reason :: any()}.
produce(Socket, Topic0, Partition, Key, Value) ->
    Topic = canonicalize_string(Topic0),
    case kafka_socket:sync(
           Socket, ?Produce, _ApiVersion = 0, _ClientID = undefined,
           {_Acks = 1,
            _Timeout = 1000,
            [{Topic, [{Partition, [{_AutoAssignOffset = -1,
                                    {_MagicByte = 1,
                                     _Attributes = 0,
                                     Key, Value}}]}]}]},
           _SockReadTimeout = infinity) of
        {ok, [{_Topic, [{Partition, ?NONE = _ErrorCode, _BaseOffset}]}]} ->
            ok;
        {ok, [{_Topic, [{Partition, ErrorCode, _}]}]} ->
            {error, ?err2atom(ErrorCode)};
        {error, _Reason} = Error ->
            Error
    end.

%% ----------------------------------------------------------------------
%% Internal functions
%% ----------------------------------------------------------------------

%% @doc Convert string to a canonic form from the point
%% of view of protocol codec.
-spec canonicalize_string(binary() | string()) -> string().
canonicalize_string(Binary) when is_binary(Binary) ->
    binary_to_list(Binary);
canonicalize_string(List) when is_list(List) ->
    List.
